src/Storages/PostgreSQL/StoragePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,514 @@
+
+    column_declaration->children.emplace_back(column_declaration->type);
+    column_declaration->children.emplace_back(column_declaration->default_expression);
+
+    return column_declaration;
+}
+
+
+ASTPtr StoragePostgreSQLReplica::getColumnDeclaration(const DataTypePtr & data_type) const
+{
+    WhichDataType which(data_type);
+
+    if (which.isNullable())
+        return makeASTFunction("Nullable", getColumnDeclaration(typeid_cast<const DataTypeNullable *>(data_type.get())->getNestedType()));
+
+    if (which.isArray())
+        return makeASTFunction("Array", getColumnDeclaration(typeid_cast<const DataTypeArray *>(data_type.get())->getNestedType()));
+
+    /// getName() for decimal returns 'Decimal(precision, scale)', will get an error with it
+    if (which.isDecimal())
+    {
+        auto make_decimal_expression = [&](std::string type_name)
+        {
+            auto ast_expression = std::make_shared<ASTFunction>();
+
+            ast_expression->name = type_name;
+            ast_expression->arguments = std::make_shared<ASTExpressionList>();
+            ast_expression->arguments->children.emplace_back(std::make_shared<ASTLiteral>(getDecimalScale(*data_type)));
+
+            return ast_expression;
+        };
+
+        if (which.isDecimal32())
+            return make_decimal_expression("Decimal32");
+
+        if (which.isDecimal64())
+            return make_decimal_expression("Decimal64");
+
+        if (which.isDecimal128())
+            return make_decimal_expression("Decimal128");
+
+        if (which.isDecimal256())
+            return make_decimal_expression("Decimal256");
+    }
+
+    return std::make_shared<ASTIdentifier>(data_type->getName());
+}
+
+
+/// For single storage PostgreSQLReplica get columns and primary key columns from storage definition.
+/// For database engine PostgreSQLReplica get columns and primary key columns by fetching from PostgreSQL, also using the same
+/// transaction with snapshot, which is used for initial tables dump.
+ASTPtr StoragePostgreSQLReplica::getCreateNestedTableQuery(const std::function<PostgreSQLTableStructure()> & fetch_table_structure)
+{
+    auto create_table_query = std::make_shared<ASTCreateQuery>();
+
+    auto table_id = getStorageID();
+    create_table_query->table = getNestedTableName();
+    create_table_query->database = table_id.database_name;
+
+    auto columns_declare_list = std::make_shared<ASTColumns>();
+    auto columns_expression_list = std::make_shared<ASTExpressionList>();
+    auto order_by_expression = std::make_shared<ASTFunction>();
+
+    auto metadata_snapshot = getInMemoryMetadataPtr();
+    const auto & columns = metadata_snapshot->getColumns();
+    NamesAndTypesList ordinary_columns_and_types;
+
+    if (!is_postgresql_replica_database)
+    {
+        ordinary_columns_and_types = columns.getOrdinary();
+    }
+    else
+    {
+        auto table_structure = fetch_table_structure();
```

Comment:
Is it possible to pass table structure from `DatabasePostgreSQLReplica` or `PostgreSQLReplicationHandler`?

Reply:
It is partially already so, now i pass lambda from `PostgreSQLReplicationHandler`, but yes, it's better to pass pointer to table structure, which will be nullptr if it is not needed to be fetched (in case it is not a postgresql replica database, but just a single storage). I'll fix.